core/src/apps/bitcoin/scripts.py

```diff
@@ -140,6 +143,22 @@ def input_script_p2pkh_or_p2sh(
     return w
 
 
+def read_input_script_p2pkh(
```

Comment:
> these functions should really be called `parse_*`, as they don't conform to the `read_*` signature

I agree.

> OTOH i'm not a big fan of the `read_*` signature; ISTM we should make a `BytearrayReader` class that keeps track of the offset internally, and exposes `read(int) -> bytes` and possibly `readinto(bytearray) -> int` methods

I was considering the same thing. What about using `uio.BytesIO`?

Reply:
At first I thought that BytesIO reallocates the data it's given, but after looking at the C source, it looks like it's using COW, see `stringio_copy_on_write()` in objstringio.c.

Reply:
Maybe worth checking out how much it enlarges the firmware size first, but it looks rather small.

also worth considering if we could use BytesIO for writing -- but it doesn't seem to provide any advantage for that

Reply:
Implemented BytearrayReader in 25e0cdac43007f33180c3d2aec91574b569dd33a. For the record, the uio module is only compiled into the emulator but not the firmware. So it's better to implement BytearrayReader ourselves than to add another module just because of BytesIO.

Related changes in: a4c9d6b2c9b68251492bab54c339743ad38992d8, b8e2bfb2a37ade221918f24c6c4d388c761824af, + I will probably also use this for parsing DER signatures